src/engine/IdTable.h

```diff
@@ -0,0 +1,933 @@
+// Copyright 2019, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@mail.uni-freiburg.de)
+
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstdlib>
+#include <cstring>
+#include <initializer_list>
+#include <ostream>
+#include "../global/Id.h"
+#include "../util/Log.h"
+
+/**
+ * @brief This struct is used to store all the template specialized data and
+ * methods of the IdTable class. This way less code of the IdTable has to
+ * be duplicated.
+ **/
+template <int COLS>
+class IdTableImpl {
+ public:
+  IdTableImpl()
+      : _data(nullptr), _size(0), _capacity(0), _manage_storage(true) {}
+  /**
+   * This is simply an Id* which has some additional methods.
+   **/
+  class ConstRow {
```

Comment:
Shouldn't this have a `Id*` field so that as a struct it does indeed have the same size as an `Id*`? Also I think it should be specified [`final`](https://en.cppreference.com/w/cpp/language/final) then we can also drop the virtual destructor

Reply:
I thought about this and then realized that the custom `Row` type for tables with a fixed with doesn't make any sense. The latest commit changes it to `std::array<Id, COLS>`, which should make the `operator[]` perform significantly more standard conform for `IdTableStatic` with fixed width.